### PR TITLE
Issue #1617: Add ability to enable/disable certain layout templates

### PR DESCRIPTION
**Pull request by docwilmot**
Fixes https://github.com/backdrop/backdrop-issues/issues/1617

### CODE REVIEW

File: core/modules/layout/config/layout.settings.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+    "_config_name": "layout.settings",
+    "templates": [
```
Comment (klonos) [inline core/modules/layout/config/layout.settings.json]: not entirely sure, but should this perhaps be `layout_allowed_templates` or `layout_enabled_templates` or something? I randomly looked at https://github.com/backdrop/backdrop/blob/1.x/core/modules/book/config/book.settings.json#L3 for example and we use `book_allowed_types` there.

File: core/modules/layout/layout.admin.inc [hunk 2]
```diff
@@ -7,6 +7,70 @@
...
+  $form['templates'] = array(
+    '#title' => t('Templates'),
+    '#type' => 'checkboxes',
+    '#default_value' => $available_layouts,
+    '#options' => array(),
+    '#wrapper_attributes' => array('class' => array('clearfix', 'layout-options')),
+  );
+  $form['show_all_admin'] = array(
+    '#title' => t('Show all for admins'),
+    '#description' => t('When enabled, all templates will be shown on layout edit pages for users with \'use all templates\' permissions.'),
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc]: We seem to be using double quotes elsewhere in help text. For example in https://github.com/backdrop/backdrop/blob/1.x/core/modules/layout/layout.admin.inc#L243:

`t('Use the "%" (percent) symbol to indicate a wildcard in the path. i.e. "node/%" or "node/%/edit".')`

File: core/modules/layout/layout.admin.inc [hunk 3]
```diff
@@ -7,6 +7,70 @@
...
+function layout_settings_page_submit($form, &$form_state) {
+  $templates = array_keys(array_filter($form_state['values']['templates']));
+  $active_templates = array();
+  $all_layouts = layout_get_all_configs('layout');
+  foreach ($all_layouts as $layout) {
+    $active_templates[] = $layout['layout'];
+  }  
+  $excluded_active_templates = array_unique(array_diff($active_templates, $templates));
+  if (!empty($excluded_active_templates)) {
+    $message = t('The following templates are in use but have been disabled: %disabled', array('%disabled' => implode(', ', $excluded_active_templates)));
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc]: This causes the layout template machine name to be used in the message shown to the users. Can we have the "pretty", admin name shown instead?

File: core/modules/layout/layout.module [hunk 4]
```diff
@@ -75,6 +75,14 @@ function layout_menu() {
     'type' => MENU_NORMAL_ITEM,
   ) + $base;
 
+  $items['admin/structure/layouts/settings'] = array(
+    'title' => 'Settings',
```
Comment (klonos) [inline core/modules/layout/layout.module]: Unless we plan to include more settings in this tab, it should be labeled something like "Manage layout templates". Better discoverability that way.

File: core/modules/layout/layout.module [hunk 5]
```diff
@@ -564,13 +572,22 @@ function layout_permission() {
       'description' => t('Add and modify layouts on pages.'),
       'restrict access' => TRUE,
     ),
+    'use all templates' => array(
+      'title' => t('Use disabled templates'),
```
Comment (klonos) [inline core/modules/layout/layout.module:576]: Include the word "layout" here (`t('Use disabled layout templates')`) and in the description below in order to not be vague (in the future we might need to refer to other types of templates in the UI). Will also help when we add a search filter in the permissions page (https://github.com/backdrop/backdrop-issues/issues/980) and people search by the term "layout".

File: core/modules/layout/layout.admin.inc (hunk 3 above)
Comment (klonos) [inline core/modules/layout/layout.admin.inc]: ...now that I think about it, using the word "disabled" here might confuse people that are used to disabling modules and how that works in that context. Perhaps use the word "hidden" here instead?

**Comment by klonos** [comment]
The "Show all for admins" checkbox does not get saved. I uncheck it, save the form and after the form is submitted, it is still checked. It is unrelated to whether there are disabled/hidden templates or not. Tried both cases and the behavior is the same.

**Comment by klonos** [comment]
Note to self: wanted to test what happens when you edit a layout after the template it uses has been disabled/hidden from the settings page, but the above glitch doesn't allow me to test properly (because as an admin, all the templates are always available).

**Comment by jenlampton** [comment]
Replaced by https://github.com/backdrop/backdrop/pull/1346

File: core/modules/layout/layout.module (hunk 5 above)
Comment (jenlampton) [inline core/modules/layout/layout.module:576]: I removed the permissions in my approach over at https://github.com/backdrop/backdrop/pull/1346 so this change was not necessary.

**Comment by jenlampton** [comment]
@klonos in my PR I added a validation check that wouldn't allow people to disable a layout that is in-use on a site.

**Comment by jenlampton** [comment]
but I'd love it if you wanted to test that one :)  I still need to write tests for it...

**Pull request closed without merging**